src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.InlineContext.cs

```diff
@@ -0,0 +1,558 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Precedence;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.InlineMethod
+{
+    internal abstract partial class AbstractInlineMethodRefactoringProvider<TInvocationSyntaxNode, TExpressionSyntax, TArgumentSyntax>
+        where TInvocationSyntaxNode : SyntaxNode
+        where TExpressionSyntax : SyntaxNode
+        where TArgumentSyntax : SyntaxNode
+    {
+        private class InlineMethodContext
+        {
+            /// <summary>
+            /// Statements should be inserted before the <see cref="StatementContainingCallee"/>.
+            /// </summary>
+            public ImmutableArray<SyntaxNode> StatementsToInsertBeforeCallee { get; }
+
+            /// <summary>
+            /// Statement invokes the callee. All the generated declarations should be put before this node.
+            /// </summary>
+            public SyntaxNode? StatementContainingCallee { get; }
+
+            /// <summary>
+            /// Inline content for the callee method. It should replace <see cref="SyntaxNodeToReplace"/>.
+            /// It will be null if nothing need to be inlined.
```

Comment:
can you calrify this?  also... when do not need to inline something... if the feature is 'inline method'? :)

Reply:
There is one case nothing will be inlined.
`void Caller{ Callee(); }`
`void Callee() { return; }`
And that's what I mean the syntax node is null here.  I can add an example here to clarify it.

Reply:
do we even want to support that case?  what does it mean to inline a void method that doesn't contain an expression/expression-statement?

Reply:
I see, then it might be better to filter this case out before the code action is provided